loader/loader_windows.c

```diff
@@ -884,37 +884,106 @@ VkResult windows_read_sorted_physical_devices(struct loader_instance *inst, uint
 
+    bool app_is_vulkan_1_1 = loader_check_version_meets_required(LOADER_VERSION_1_1_0, inst->app_api_version);
+
+    // Iterate over all devices and make sure any ICD's with matching LUID's are sorted such that drivers with a underlyingAPI of
+    // VK_LAYERED_DRIVER_UNDERLYING_API_D3D12_MSFT are ordered after drivers without it
+    struct loader_phys_dev_per_icd *sorted_array = *sorted_devices;
+    for (uint32_t i = 0; *sorted_devices_count > 1 && i < *sorted_devices_count - 1; i++) {
+        for (uint32_t j = i + 1; j < *sorted_devices_count; j++) {
+            // Skip consideration for reordering if LUID's dont match
+            if ((sorted_array[i].luid.HighPart != sorted_array[j].luid.HighPart) &&
+                (sorted_array[i].luid.LowPart != sorted_array[j].luid.LowPart)) {
+                continue;
+            }
+
+            VkLayeredDriverUnderlyingApiMSFT underlyingAPI = VK_LAYERED_DRIVER_UNDERLYING_API_NONE_MSFT;
+            VkPhysicalDeviceLayeredDriverPropertiesMSFT layered_driver_properties_msft = {0};
+            layered_driver_properties_msft.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_LAYERED_DRIVER_PROPERTIES_MSFT;
+            VkPhysicalDeviceProperties2 props2 = {0};
+            props2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2;
+            props2.pNext = (void *)&layered_driver_properties_msft;
+
+            for (uint32_t k = 0; k < sorted_array[i].device_count; k++) {
+                VkPhysicalDeviceProperties dev_props = {0};
+                sorted_array[i].icd_term->dispatch.GetPhysicalDeviceProperties(sorted_array[i].physical_devices[k], &dev_props);
+
+                bool device_is_1_1_capable =
+                    loader_check_version_meets_required(LOADER_VERSION_1_1_0, loader_make_version(dev_props.apiVersion));
+
+                PFN_vkGetPhysicalDeviceProperties2 GetPhysDevProps2 = NULL;
+                if (app_is_vulkan_1_1 && device_is_1_1_capable) {
+                    GetPhysDevProps2 = sorted_array[i].icd_term->dispatch.GetPhysicalDeviceProperties2;
+                } else {
+                    GetPhysDevProps2 =
+                        (PFN_vkGetPhysicalDeviceProperties2)sorted_array[i].icd_term->dispatch.GetPhysicalDeviceProperties2KHR;
+                }
+                if (GetPhysDevProps2) {
+                    GetPhysDevProps2(sorted_array[i].physical_devices[k], &props2);
+                    if (layered_driver_properties_msft.underlyingAPI != VK_LAYERED_DRIVER_UNDERLYING_API_NONE_MSFT) {
+                        underlyingAPI = layered_driver_properties_msft.underlyingAPI;
+                        break;
+                    }
+                }
+            }
+            if (underlyingAPI == VK_LAYERED_DRIVER_UNDERLYING_API_D3D12_MSFT) {
+                struct loader_phys_dev_per_icd swap_icd = sorted_array[i];
+                sorted_array[i] = sorted_array[j];
+                sorted_array[j] = swap_icd;
+            }
+        }
+    }
+
```

Comment:
Overall, this function is too large for someone new to the code to quickly build an accurate picture of what's going on. Several pieces could benefit from being hoisted into helper functions that have descriptive function names.  Some of the nested scopes could benefit from refactoring in the manner of "never nester".

Reply:
I hadn't heard about "never nester" but already strongly identify with it!